### ISSUE TITLE
Logout on Inventory immediately after login

### ISSUE THREAD

**Issue by adborden** [GSA/data.gov#2036]
This is a longstanding issue. I often have to login several times before I can accomplish my task.

This can be extremely frustrating for users, especially when they are in the middle of a task, like filling out the new dataset form only to be logged out when they got submit and lose all their work.

## How to reproduce

In an incognito window:
1. inventory.data.gov/user/login
1. Login with MAX.gov
1. Click Datasets
1. Click Add New Dataset
1. Fill in info

The issue is intermittent. Generally,clicking around will log you out.

## Expected behavior

I login once and I'm logged in until I'm inactive for X minutes as per SSP.

## Actual behavior
 After a few clicks I. Logged out, I see the saml unauthorized screen and have to log in again.

**Comment by adborden** [comment]
I'm pretty sure this is related to ckanext-saml2 and how sessions are handled. Unlike catalog, we have multiple instances handling sessions. It seems like you might get a session from one instance but if your next request hits the other instance, you'll be logged out. You have to login several times before both instances are seeing a valid session.

We know that pysaml2 stores sessions in memory or file, which is not cloud friendly. Our fork of pysaml2 includes a memcached store, but even of that is correct, each instance is using its own memcached instance, defeating the purpose. Still, I'm not sure if this store is only used in the saml2 authentication or if it's also used beyond that for sessions.

Beaker, CKAN's session store is configured to store sessions in the database.

**Comment by FuhuXia** [comment]
Tested on staging. Tried to get authenticated on one instance then access the other instance by toggling instance's apache2 on and off. Could not replicate the issue. Looks like session is valid across instances, and stored in DB, not memcached on instance.

Issue is observed on live production. Will do some further testing during off hours.

From the info I gathered so far, a wild guess is that is might have something to do with short-lived cached sql query in beaker. If that is case, it might have been fixed in CKAN 2.8. In the meanwhile, if we can tell NetScaler to increase session stickiness in the Round Robin Load Balancing, that will help to alleviate the issue.
